### ISSUE TITLE
Editor: Console Output Advanced Boolean Filters

### ISSUE THREAD

**Issue by RobertBColton** [godotengine/godot-proposals#11244]
### Describe the project you are working on

ALL my projects are affected, it's the editor usability for me to navigate the output.

### Describe the problem or limitation you are having in your project

It's not specific to any one project

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

This is a suggestion to add boolean logic to the filter bar on the output console. This is a really important feature in other editors, as I want to exclude messages as well as include them. I don't have a specific syntax in mind (could be regexp) but I suggest using the GitHub one maybe?

In VS Code the console just has a regular find/replace Window, but I would think that would be overkill. I wouldn't do that in Godot since Godot should be more beginner friendly.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

You could use boolean operations and categories in the filter bar, like you can when searching on GitHub.

![Image](https://github.com/user-attachments/assets/a0d3a622-2b0e-47b0-9cad-57ab0d8bb337)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, this feature would be used often

### Is there a reason why this should be core and not an add-on in the asset library?

It's kind of essential for debugging to find specific messages. Log-level splitting doesn't go far enough as you can potentially have many messages per log level.

**Comment by FilipeAlexCosta** [comment]
Hi! Just a quick note that @rafaelgcruz and I would like to play around with this idea, if no one objects.

I've seen that #11850 mentions adding regex support to the filtering and we believe this can be integrated with Robert's github-style syntax idea, while also providing a friendlier interface to those less familiar with regex.

I do think the syntax itself would still need to be worked out. We were thinking something like ' ' (space) being an implicit AND between two expressions (like Go dot meaning the log must include "Go" AND "dot") and 'OR' for alternatives, much like in github. For excluding we were thinking on starting with 'NOT' instead of 'exclude:', but perhaps the latter is better?

Thanks!

**Comment by Calinou** [comment]
> We were thinking something like ' ' (space) being an implicit AND between two expressions

This means you wouldn't be able to look for an exact substring that contains spaces unless you quote it though.

**Comment by FilipeAlexCosta** [comment]
> This means you wouldn't be able to look for an exact substring that contains spaces unless you quote it though.

That's true. I feel like it would be fine in most scenarios (I'm obviously biased though since I suggested it), but I can see it becoming pretty annoying in others.

In fact, now that I think about it, spaces for ANDs could also be problematic if implemented in an unordered way, that is, searching for `firstTerm secondTerm` would match those two terms in any order, which would be even more confusing, especially for users used to the current interface. I'll spend some more time thinking about it.

I like the hyphen idea!

**Comment by rafaelgcruz** [comment]
Hi! So, as promised, over the last week we've polished our implementation and we believe it has reached a point where, to improve it further, feedback from more experienced Godot developers is needed (especially since it's our first time trying to implement a feature, meaning there will probably be a few kinks to iron out).

The parser and filter itself seem to be working correctly, nevertheless, we'd still like to write some unit tests if possible. However, we couldn't find any dedicated folder for editor-related tests. Is it normal not to test editor functionality directly, or are we missing something?

Also, we've seen that it's not unusual for contributors working on new features to open draft PRs in the godot repo, should we also do that or is a "normal" PR fine?

If possible, we'd appreciate some help regarding these two questions @Calinou :)

Thanks!

**Comment by Calinou** [comment]
> The parser and filter itself seem to be working correctly, nevertheless, we'd still like to write some unit tests if possible. However, we couldn't find any dedicated folder for editor-related tests. Is it normal not to test editor functionality directly, or are we missing something?

We don't have the ability to test editor-only code yet, so you don't need to add unit tests yet.

> Also, we've seen that it's not unusual for contributors working on new features to open draft PRs in the godot repo, should we also do that or is a "normal" PR fine?

Feel free to open a draft PR 🙂
